File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.go
import "math"

func minSubArrayLen(target int, nums []int) int {
    leftIndex := 0
    sum := 0; 
    minLength := math.MaxInt32 

    for rightIndex := 0 ; rightIndex < len(nums) ; rightIndex++ {
        sum += nums[rightIndex];

        //여기서 왼쪽인덱스를 줄여나간다. 
        for sum >= target {
            minLength = min(minLength, rightIndex-leftIndex + 1)
            sum -= nums[leftIndex]
            leftIndex++
        }

    }

    if minLength == math.MaxInt32 {
        return 0;
    }

    return minLength;    
}


func min(a,b int) int {
    if a < b {
        return a
    }

    return b
}

//먼저 오른쪽 맨끝을 정할 right를 루프를 돌아서 정함

//이중 반복문으로 해당 right에서 왼쪽을 줄여가보면서 target에 맞는지 정한다. 

// 가장 짧은 길이의 배열을 구하는 것이니 minLength 는 Math.int32

//두 값을 비교하는건 math.Min

File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(target, nums) {
    //하나의 포인터를 두고 왼쪽에서 줄여 나가는 식으로 해야함

    // 포인터에서 왼쪽 포인터를 두고 줄이는 방법은 O(n  * n ) O(n^2) 으로 푸는 방법
    let leftIndex = 0;
    let sum = 0;
    let minLen = Infinity;

    for(let idx = 0 ; idx < nums.length ; idx++){
        sum += nums[idx];

        while(sum >= target){
            minLen = Math.min(minLen, idx - leftIndex + 1);
            sum -= nums[leftIndex];
            leftIndex++;
        }
    }

    if (minLen == Infinity) return 0;
    

    return minLen

};

// 양수의 숫자만 있는 경우 target과 같은 가장 적은 길이를 가진 subArray를 가져야함
//슬라이딩 윈도우 기법
